### PR TITLE
UF-5912: Add paging model

### PR DESCRIPTION
**Pull request by ronnberg**

### CODE REVIEW

File: dept44-api-models/src/main/java/se/sundsvall/dept44/api/models/paging/PagingMetaData.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+	@Schema(description = "Displayed objects on current page", example = "13", accessMode = READ_ONLY)
+	private int count;
+
+	@Schema(description = "Total amount of hits based on provided search parameters", example = "98", accessMode = READ_ONLY)
+	private long totalRecords;
+
+	@Schema(description = "Total amount of pages based on provided search parameters", example = "23", accessMode = READ_ONLY)
+	private int totalPages;
+
+	@Schema(description = "The properties to sort by", example = "property", accessMode = READ_ONLY)
```
Comment (mha109) [inline dept44-api-models/src/main/java/se/sundsvall/dept44/api/models/paging/PagingMetaData.java:42]: `@ArraySchema` here as well?

File: dept44-api-models/src/main/java/se/sundsvall/dept44/api/models/paging/validation/impl/MaxPagingLimitConstraintValidator.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+public class MaxPagingLimitConstraintValidator implements ConstraintValidator<MaxPagingLimit, Integer> {
+
+	private static final String MSG = "Page limit cannot be greater than %s";
+
+	@Value("${dept44.api.models.paging.max.limit:100}")
+	private int maxLimit;
+
+	@Override
+	public boolean isValid(Integer value, ConstraintValidatorContext context) {
+		var isValid =  value <= maxLimit;
```
Comment (mha109) [inline dept44-api-models/src/main/java/se/sundsvall/dept44/api/models/paging/validation/impl/MaxPagingLimitConstraintValidator.java:17]: Captain OCD here. two spaces after the `=` ;)

File: dept44-api-models/src/main/java/se/sundsvall/dept44/api/models/paging/PagingMetaData.java [hunk 3]
```diff
@@ -42,7 +43,7 @@ public class PagingMetaData {
 	@Schema(description = "The properties to sort by", example = "property", accessMode = READ_ONLY)
 	private List<String> sortBy;
 
-	@Schema(description = "The sort order direction", example = "ASC", enumAsRef = true)
+	@ArraySchema(schema = @Schema(description = "The sort order direction", example = "ASC", enumAsRef = true))
```
Comment (foobar-1337) [inline dept44-api-models/src/main/java/se/sundsvall/dept44/api/models/paging/PagingMetaData.java:46]: Shouldn't `@ArraySchema` be placed on sortBy-variable instead?

('sortBy' is a List, 'sortDirection' is not a List)

File: dept44-api-models/pom.xml [hunk 4]
```diff
@@ -0,0 +1,95 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+    <parent>
+        <groupId>se.sundsvall.dept44</groupId>
+        <artifactId>dept44-parent</artifactId>
+        <version>3.1-SNAPSHOT</version>
+    </parent>
+    <artifactId>dept44-api-models</artifactId>
```
Comment (foobar-1337) [inline dept44-api-models/pom.xml:9]: Maybe rename to `dept44-models`? 
It's a more generic name and not restricted to pojos för the api-layer.

File: dept44-api-models/src/main/java/se/sundsvall/dept44/api/models/paging/validation/impl/ValidSortByPropertyConstraintValidator.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+import java.lang.reflect.Field;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Stream;
+
+public class ValidSortByPropertyConstraintValidator implements ConstraintValidator<ValidSortByProperty, AbstractParameterPagingBase> {
+
+	private static final String CUSTOM_ERROR_MESSAGE_TEMPLATE = "One or more of the sortBy members %s are not valid. Valid properties to sort by are %s.";
```
Comment (foobar-1337) [inline dept44-api-models/src/main/java/se/sundsvall/dept44/api/models/paging/validation/impl/ValidSortByPropertyConstraintValidator.java:20]: "One or more of the sortBy members %s are not valid...."
->
"One or more of the sortBy properties %s are not valid...."

File: dept44-api-models/pom.xml [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+            <artifactId>dept44-starter</artifactId>
+            <scope>provided</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-starter-data-jpa</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.projectlombok</groupId>
+            <artifactId>lombok</artifactId>
```
Comment (foobar-1337) [inline dept44-api-models/pom.xml:24]: Personal opinion:

There are not a huge amount of properties in the model classes. 
The need of an extra dependency for the one-time job of generating gettters/setters can be questioned.

Comment (ronnberg) [reply dept44-api-models/pom.xml]: Noted

File: dept44-models/src/main/java/se/sundsvall/dept44/models/api/paging/AbstractParameterPagingBase.java [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+@EqualsAndHashCode
+@Getter
+@Setter
+public abstract class AbstractParameterPagingBase {
+
+	private static final String DEFAULT_PAGE = "1";
+	private static final String DEFAULT_LIMIT = "100";
+
+
+	@Schema(description = "Page number", example = "1")
```
Comment (jklingst) [inline dept44-models/src/main/java/se/sundsvall/dept44/models/api/paging/AbstractParameterPagingBase.java:34]: For extra clarity, maybe add something about the page numbering being 1-based to the schema description? Yes, the example states one, but still

File: dept44-models/pom.xml [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+            <scope>provided</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-starter-data-jpa</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.projectlombok</groupId>
+            <artifactId>lombok</artifactId>
+            <version>1.18.30</version>
```
[the remark this answers is not in the document]
Comment (ronnberg) [reply dept44-models/pom.xml]: Removed

File: dept44-models/src/main/java/se/sundsvall/dept44/models/api/paging/AbstractParameterPagingBase.java (hunk 7 above)
Comment (ronnberg) [reply dept44-models/src/main/java/se/sundsvall/dept44/models/api/paging/AbstractParameterPagingBase.java]: Minimum added

**Review by mha109 — APPROVED** [review]

**Review by mha109 — APPROVED** [review]

**Review by jklingst — APPROVED** [review]

**Review by josc04 — APPROVED** [review]

File: dept44-models/src/main/java/se/sundsvall/dept44/models/api/paging/validation/impl/ValidSortByPropertyConstraintValidator.java [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+			.filter(name -> Arrays.stream(constraintAnnotation.exclude()).noneMatch(name::equals))
+			.collect(collectingAndThen(toList(), list -> {
+				list.addAll(Arrays.asList(constraintAnnotation.include()));
+				return list;
+			})));
+	}
+
+	@Override
+	public boolean isValid(final AbstractParameterPagingAndSortingBase parameters, final ConstraintValidatorContext context) {
+		final boolean isValid = isEmpty(parameters.getSortBy()) ||	entityProperties.containsAll(parameters.getSortBy());
```
Comment (josc04) [inline dept44-models/src/main/java/se/sundsvall/dept44/models/api/paging/validation/impl/ValidSortByPropertyConstraintValidator.java:41]: Assistant to Captain OCD says to switch from tab to space after `||`.

**Review by josc04 — APPROVED** [review]

**Review by mha109 — APPROVED** [review]

**Pull request merged**